lib/http2/hpack.c

```diff
@@ -306,11 +376,23 @@ static int decode_header(h2o_mem_pool_t *pool, struct st_h2o_decode_header_resul
     } else {
         /* non-existing name */
         const h2o_token_t *name_token;
-        if ((result->name = decode_string(pool, src, src_end)) == NULL)
+        uint8_t flags = 0;
+        if ((result->name = decode_string(pool, src, src_end, &flags)) == NULL)
             return H2O_HTTP2_ERROR_COMPRESSION;
-        if (contains_uppercase(result->name->base, result->name->len)) {
-            *err_desc = "found an upper-case letter in header name";
-            return H2O_HTTP2_ERROR_PROTOCOL;
+        /* for pseudo-headers, the validation is done in decode_header,
+           only deal with the other headers */
+        if (result->name->base[0] != ':') {
+            if (flags & H2O_HUFF_ENCODED) {
+                if (flags & H2O_HUFF_INVALID_FOR_HEADER_NAME) {
+                    *err_desc = "found an invalid character in header name";
+                    return H2O_HTTP2_ERROR_PROTOCOL;
```

Comment:
How about using HTTP response level error notification for invalid header names as well?